### ISSUE TITLE
Some users need to unblock Modnix to be able to see/edit Configuration

### ISSUE THREAD

**Issue by mad2342** [Sheep-y/Modnix#33]
I use the excellent option to have a class in my Assembly that Modnix parses to provide the configuration view for editable settings.
I had many posts on nexus on how to configure "Assorted Adjustments", it seems that quite a lot of users need to "unblock" Modnix for the Config-Tab to be shown at all.
It's fixable on the users side following this guide: https://stackoverflow.com/questions/28339116/not-allowed-to-load-assembly-from-network-location
but if possible would be better to workaround/fix on programs side as fiddling with .NET settings isn't really reliable for the end-users.
I can't provide more details, unfortunately, as i never had the issue myself. I'll point people to this issue though, if you're fine with that?
Thx alot and best regards, Mad

**Comment by mad2342** [comment]
New info: A user reported that a mod of yours "Independent Gear" shows the config just fine, while mine won't.
I had a look at the changes and your impentation of loading the settings is different indeed.
I followed your guide in the WIKI and my stuff looks like this:
`
namespace AssortedAdjustments
{
    public static class AssortedAdjustments
    {
        internal static Settings Settings;

        // Modnix Entrypoints
        public static void SplashMod(Func<string, object, object> api)
        {
            Settings = api("config", null) as Settings ?? new Settings();
            (...)
`
with Settings:
`
namespace AssortedAdjustments
{
    internal class Settings
    (...)
`
Only potential problem could be the different entrypoint? (Splash vs. Main)?

**Comment by Sheep-y** [comment]
I read on Discord that the second config issue has been solved, so I'll focus on the first.

In order to properly load a mod's config class, Modnix need to create a .Net sandbox to load the mod.
This sandbox protects Modnix from crash, and keep the mod dll free from file lock so that it is updatable, deletable etc.
The `loadFromRemoteSources` config affects this sandbox.

As of Modnix 2.5.4 and Modnix 3 Beta 1, the config file will be created when the sandbox [failed to initiate](https://github.com/Sheep-y/Modnix/blob/v2.5.4/MainGUI/ModLoaderBridge.cs#L654).
The unfortunately news is I can't reproduce it.  I tried marking installer and installed Modnix as Downloaded from Internet, and it still runs ok after the unlock prompt.  I also tried marking the  the mod, which create another Exception that is not solved by the `loadFromRemoteSources` config.

So, what I can do now, is to create the config anyway during startup, and add it to manual install package.  Also Modnix can check on launch whether the config exists and, if not, create it and restart (if created).  This should make sure Modnix is under the effect of loadFromRemoteSources.

As for the exception when the mod is marked as downloaded, for now the workaround is to remove the mark from the dll.  Usually done by right-clicking it and select properties, but the Add Mod button also works.

I am seeing how Modnix can remove the mark.  For reference, this is the error I got in the config panel:

```
System.IO.FileLoadException: Could not load file or assembly 'file:///C:\Users\Sheepy\Documents\My Games\Phoenix Point\Mods\PhoenixPt_DumpData\PhoenixPt_DumpData.dll' or one of its dependencies. Operation is not supported. (Exception from HRESULT: 0x80131515)
File name: 'file:///C:\Users\Sheepy\Documents\My Games\Phoenix Point\Mods\PhoenixPt_DumpData\PhoenixPt_DumpData.dll' ---> System.NotSupportedException: An attempt was made to load an assembly from a network location which would have caused the assembly to be sandboxed in previous versions of the .NET Framework. This release of the .NET Framework does not enable CAS policy by default, so this load may be dangerous. If this load is not intended to sandbox the assembly, please enable the loadFromRemoteSources switch. See http://go.microsoft.com/fwlink/?LinkId=155569 for more information.
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at Sheepy.Modnix.MainGUI.Sandbox.LoadDll(String path)
```

**Comment by Sheep-y** [comment]
Here is a Modnix 3 development build that implements all the changes I laid above.
From my testing it is able to unblock and read/write config of web-blocked mods.

[ModnixInstaller_v3_20210111.zip](https://github.com/Sheep-y/Modnix/files/5793920/ModnixInstaller_v3_20210111.zip)

**Comment by Sheep-y** [comment]
Zyxpsilon and mad2342 confirmed that the build is working.  The changes have been backported and released with [v2.5.5](https://github.com/Sheep-y/Modnix/releases/tag/v2.5.5).

**Comment by pleszczy** [comment]
Tried it on [v2.5.6](https://github.com/Sheep-y/Modnix/releases/tag/v2.5.6) and the same thing with 'This release of the .NET Framework does not enable CAS policy' happened.

It only started working after following that configuration suggestion from stack overflow.
``` xml
<configuration>
    <runtime>
       <loadFromRemoteSources enabled="true"/>
   </runtime>
</configuration>
```

**Comment by Sheep-y** [comment]
@pleszczy Hi. Sorry to hear that it is still an issue.  May I know:

1. Was your change made to machine.config in the .net config folder?
2. Was the Modnix setup / update done automatically or manually?
3. Does a `Modnix.conf` file exists in the same folder as `Modnix.exe`
4. If yes, does it have the same content (`loadFromRemoteResource`) and does it have the same name as the exe?

If question 3 or 4 is not, i.e. if the conf file does not exist or it is empty, it would be nice to have the setup log too.

**Comment by pleszczy** [comment]
> @pleszczy Hi. Sorry to hear that it is still an issue. May I know:
> 
> 1. Was your change made to machine.config in the .net config folder?
> 2. Was the Modnix setup / update done automatically or manually?
> 3. Does a `Modnix.conf` file exists in the same folder as `Modnix.exe`
> 4. If yes, does it have the same content (`loadFromRemoteResource`) and does it have the same name as the exe?
> 
> If the answer to question 3 or 4 is no, i.e. if the conf file does not exist or it is empty, it would be nice to have the setup log too.

I've started from a scratch and made screenshot of every step I've made to help you reproduce it.

**Steps to reproduce**
- Download the latest dev version of Modnix (just to check if its maybe fixed already) [v3.0.2021.0204](https://github.com/Sheep-y/Modnix/releases/tag/v3.0.2021.0204)
- Install Modnix as an administrator
<img width="625" alt="1" src="https://user-images.githubusercontent.com/476428/107386446-288fa880-6af4-11eb-8cc3-91ea5852326b.PNG">
- Download the latest version of  AssortedAdjustments [1.9.3.1](https://github.com/Mad-Mods-Phoenix-Point/AssortedAdjustments/releases/tag/1.9.3.1)
- Installed it by using the 'add mod' button
<img width="1190" alt="2" src="https://user-images.githubusercontent.com/476428/107387060-be2b3800-6af4-11eb-9290-413304250171.PNG">
- Selected configuration in AssortedAdjustments
<img width="1347" alt="3" src="https://user-images.githubusercontent.com/476428/107387219-e5820500-6af4-11eb-8a59-d92c0504ead3.PNG">
- `Modnix.conf`
``` json
{
  "SettingVersion": 20200428,
  "LogLevel": 15,
  "LogMonitor": false,
  "CheckUpdate": true,
  "LastCheckUpdate": "2021-02-09T16:20:28.2183637+01:00",
  "UpdateChannel": "dev",
  "GamePath": "F:\\SteamLibrary\\steamapps\\common\\Phoenix Point",
  "ModInfoWeight": -1.0,
  "ModListWeight": -1.0,
  "MinifyLoaderPanel": false,
  "MinifyGamePanel": true,
  "MaximiseWindow": false,
  "WindowLeft": -1.0,
  "WindowTop": -1.0,
  "WindowWidth": -1.0,
  "WindowHeight": -1.0,
  "OfflineParameter": "",
  "SteamCommand": "steam://rungameid/839770",
  "GogExe": null,
  "GogParameter": "/gameId=1795581746 /command=runGame /path=\"%GAME_PATH%\"",
  "EgsCommand": "com.epicgames.launcher://apps/Iris?action=launch",
  "EgsParameter": "",
  "Mods": null
}
```
- `Modnix.config`
``` xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
	<runtime>
		<loadFromRemoteSources enabled="true"/>
	</runtime>
</configuration>
```

**Comment by pleszczy** [comment]
**Steps to fix it**

- find machine.config in `%windir%\Microsoft.NET\Framework64\[version]\config\machine.config`

<img width="400" alt="4" src="https://user-images.githubusercontent.com/476428/107390428-fd0ebd00-6af7-11eb-8f8b-956b7dc8c512.PNG">

- add loadFromRemoteSources

``` xml
<runtime>
       <loadFromRemoteSources enabled="true"/>
</runtime>
```

<img width="400" alt="5" src="https://user-images.githubusercontent.com/476428/107390583-20396c80-6af8-11eb-8f05-bbae3c5ed07e.PNG">

<img width="400" alt="6" src="https://user-images.githubusercontent.com/476428/107390842-62fb4480-6af8-11eb-9286-ac1ce5bde847.PNG">

- config works

<img width="400" alt="7" src="https://user-images.githubusercontent.com/476428/107391270-d2713400-6af8-11eb-8bff-30bbeeebd69e.PNG">

Though this has a side effect of breaking nordVPN and openHardwareMonitor. theyll complain that there is an unknown configuration present in the machin config

**Comment by Sheep-y** [comment]
Thanks pleszczy, that is very detailed.  And thanks for the English message.

The reason I asked the info is because the last time I could reproduce it, it could be fixed with the local config, thus I consider it fixed.  Apparently not for everyone.

I definitely can't reproduce it on my development and gaming pc, hard as I try, so I'll see if I can reproduce it on the cleanest pc in my home.

**Comment by Sheep-y** [comment]
Happy Chinese New Year!  I am back!  Who wants red pockets?

I tried running Modnix 3 Beta 3 on a "clean" media/gaming machine on an unprivileged account, with my own AA build, still can't reproduce the error.  Will try a formal release next and, failing that, a totally clean Win 7 VM.

I start to think it may be related with Windows languages because I remember the previous config issue was also reported on a non-English platform.  In that case it may be more difficult to test.  We'll see.

**Comment by Sheep-y** [comment]
Did all I can.  Create a new [Win 10 1809 VM](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/), make it Japanese, downloaded [Modnix](https://github.com/Sheep-y/Modnix/releases/tag/v3.0.2021.0204) and [Assorted Adjustment ](https://github.com/Mad-Mods-Phoenix-Point/AssortedAdjustments/releases/tag/1.10.0.2).

Modnix will fail to enable Add Mod button on first launch, which will be fixed.  The Add Mod button is enabled on subsequence launch, and can install the zip, and can show the config.

I'll try a little bit more, but unless there is a breakthrough I'll stop the report here.  Consider this issue "Cannot Reproduce", but I'm leaving this open for visibility.

**Comment by Aghiuta2** [comment]
Had to make an account to add my two cents. All the program editing and .NET and all that went a bit above my head, while trying to solve this issue. But since my 286 days I know that sometimes just checking all tabs/files/setting, might lead to a solution.
So I found that if you right click on the Modnix.exe file -> select Properties -> at the very bottom (check photo) there will be a button labeled Unblock, click that and after that the installed Modnix will allow proper viewing of the config section.
![image](https://user-images.githubusercontent.com/85261840/120563393-20e07080-c411-11eb-811b-49288a004f4e.png)
